simulator/simulator.go

```diff
@@ -0,0 +1,329 @@
+		return errors.Wrap(err, "transport error publishing logs")
+	}
+	if invalid {
+		return errors.New("enrollment invalid in publish logs")
+	}
+	return nil
+}
+
+// RequestQueries is the implementation of the host simulation's request new
+// queries functionality
+func (h *HostSimulation) RequestQueries() error {
+	h.state.lock.Lock()
+	defer h.state.lock.Unlock()
+
+	queries, invalid, err := h.state.serviceClient.RequestQueries(context.Background(), h.state.nodeKey)
+	if err != nil {
+		return errors.Wrap(err, "transport error requesting queries")
+	}
+	if invalid {
+		return errors.New("enrollment invalid in request queries")
+	}
+
+	if len(queries.Queries) == 0 {
+		// No queries to run
+		return nil
+	}
+
+	results := []distributed.Result{}
+	for name, sql := range queries.Queries {
+		rows, err := h.host.RunQuery(sql)
+		if err != nil {
+			rows = []map[string]string{}
+		}
+		results = append(results,
+			distributed.Result{QueryName: name, Status: 0, Rows: rows},
+		)
+	}
+
+	_, _, invalid, err = h.state.serviceClient.PublishResults(context.Background(), h.state.nodeKey, results)
+	if err != nil {
+		return errors.Wrap(err, "transport error publishing results")
+	}
+	if invalid {
+		return errors.New("enrollment invalid in publish results")
+	}
+
+	return nil
+}
+
+// hostSimulationState is a light container around simulation state management
+type hostSimulationState struct {
+	lock          sync.RWMutex
+	serviceClient service.KolideService
+	nodeKey       string
+	failed        bool
+	started       bool
+}
+
+// SimulationOption is a functional option pattern for defining how a host
+// simulation instance should be configured. For more information on this
+// patter, see the following blog post:
+// https://dave.cheney.net/2014/10/17/functional-options-for-friendly-apis
+type SimulationOption func(*HostSimulation)
+
+// WithRequestQueriesInterval is a functional option which allows the user to
+// specify how often the simulation host should check-in to the server and ask
+// for queries to run
+func WithRequestQueriesInterval(interval time.Duration) SimulationOption {
+	return func(i *HostSimulation) {
+		i.requestQueriesInterval = interval
+	}
+}
+
+// WithRequestConfigInterval is a functional option which allows the user to
+// specify how often the simulation host should check-in to the server and ask
+// for a new config
+func WithRequestConfigInterval(interval time.Duration) SimulationOption {
+	return func(i *HostSimulation) {
+		i.requestConfigInterval = interval
+	}
+}
+
+// WithPublishLogsInterval is a functional option which allows the user to
+// specify how often the simulation host should log status and results logs
+func WithPublishLogsInterval(interval time.Duration) SimulationOption {
+	return func(i *HostSimulation) {
+		i.publishLogsInterval = interval
+	}
+}
+
+// WithInsecure is a functional option which allows the user to declare that the
+// remote API should be connected to over an insecure channel
```

Comment:
This comment (and the one on line 199) don't really accurately explain what these options do. IIRC `WithInsecure()` will skip cert verification but still use TLS, while `WithInsecureGrpc()` will turn off TLS entirely.